yarn-project/sequencer-client/src/sequencer/public_processor.ts

```diff
@@ -73,65 +91,133 @@ export class PublicProcessor {
+    const outputRoots = output.constants.historicTreeRoots.privateHistoricTreeRoots;
+    outputRoots.nullifierTreeRoot = Fr.fromBuffer(nullifierTreeInfo.root);
+    outputRoots.contractTreeRoot = Fr.fromBuffer(contractTreeInfo.root);
+    outputRoots.privateDataTreeRoot = Fr.fromBuffer(privateDataTreeInfo.root);
 
-    publicKernelOutput.constants.historicTreeRoots.privateHistoricTreeRoots.nullifierTreeRoot = Fr.fromBuffer(
-      nullifierTreeInfo.root,
-    );
-    publicKernelOutput.constants.historicTreeRoots.privateHistoricTreeRoots.contractTreeRoot = Fr.fromBuffer(
-      contractTreeInfo.root,
-    );
-    publicKernelOutput.constants.historicTreeRoots.privateHistoricTreeRoots.privateDataTreeRoot = Fr.fromBuffer(
-      privateDataTreeInfo.root,
-    );
-
-    return [publicKernelOutput, publicKernelProof];
+    const proof = await this.publicProver.getPublicKernelCircuitProof(output);
+    return [output, proof];
   }
 
-  protected getPublicCallData(
-    txRequest: TxRequest,
-    bytecodeHash: Fr,
-    publicCircuitOutput: PublicCircuitPublicInputs,
-    publicCircuitProof: Proof,
-  ) {
-    // The first call is built from the tx request directly with an empty stack
-    const contractAddress = txRequest.to;
-    const callStackItem = new PublicCallStackItem(contractAddress, txRequest.functionData, publicCircuitOutput);
-    const publicCallStackPreimages: PublicCallStackItem[] = times(PUBLIC_CALL_STACK_LENGTH, PublicCallStackItem.empty);
-    const portalContractAddress = publicCircuitOutput.callContext.portalContractAddress.toField();
-
-    // set the msgSender for each call in the call stack
-    for (let i = 0; i < publicCallStackPreimages.length; i++) {
-      const isDelegateCall = publicCallStackPreimages[i].publicInputs.callContext.isDelegateCall;
-      publicCallStackPreimages[i].publicInputs.callContext.msgSender = isDelegateCall
-        ? callStackItem.publicInputs.callContext.msgSender
-        : callStackItem.contractAddress;
+  protected getKernelCircuitOutput(
+    txRequest: SignedTxRequest,
+    callData: PublicCallData,
+    previousOutput: KernelCircuitPublicInputs | undefined,
+    previousProof: Proof | undefined,
+  ): Promise<KernelCircuitPublicInputs> {
+    if (previousOutput && previousProof) {
+      if (previousOutput.isPrivateKernel) {
+        throw new Error(`Calling public functions from private ones is not implemented yet`);
+      }
+
+      const vk = getVerificationKeys().publicKernelCircuit;
+      const vkIndex = 0;
+      const vkSiblingPath = MembershipWitness.random(VK_TREE_HEIGHT).siblingPath;
+      const previousKernel = new PreviousKernelData(previousOutput, previousProof, vk, vkIndex, vkSiblingPath);
+      const inputs = new PublicKernelInputs(previousKernel, callData);
+      // TODO: This should be set by the public kernel circuit
+      inputs.previousKernel.publicInputs.end.publicCallCount = new Fr(1n);
```

Comment:
@PhilWindle see https://github.com/AztecProtocol/aztec-packages/issues/487

Reply:
Yeah, in short I didn't have the public circuit update this because the private circuit doesn't update the private equivalent. I don't know if this should be updated by the circuit or merely validated. I can't see anywhere that updates the private call count during private execution, either in the circuit or outside.

Reply:
Straightforward change to make once we know where to make it @iAmMichaelConnor